### PR TITLE
Docs: update description of visible_hostname

### PR DESCRIPTION
**Pull request by yadij**
Bring the documentation back into line with what the code
actually does.

Also, mention that this is an domain/FQDN not just a host name.

### CODE REVIEW

File: src/cf.data.pre [hunk 1]
```diff
@@ -7112,11 +7112,22 @@ LOC: Config.visibleHostname
 DEFAULT: none
 DEFAULT_DOC: Automatically detect the system host name
 DOC_START
-	If you want to present a special hostname in error messages, etc,
-	define this.  Otherwise, the return value of gethostname()
-	will be used. If you have multiple caches in a cluster and
-	get errors about IP-forwarding you must set them to have individual
-	names with this setting.
+	The public domain name (FQDN) of this Proxy.
```
Comment (rousskov) [inline src/cf.data.pre]: I do not think this name has to be "public" or even fully qualified. In many deployments, it is and will be neither.

I recommend describing what this name is used for without unnecessarily (or misleadingly) restricting its scope.

Comment (rousskov) [inline src/cf.data.pre]: Please do not capitalize Proxy. It looks strange, grammatically incorrect, and/or inconsistent.

File: src/cf.data.pre [hunk 2]
```diff
@@ -7112,11 +7112,22 @@ LOC: Config.visibleHostname
...
+
+	This name will be used in generated URLs sent to clients for
+	objects they need to download directly from Squid, and for other
+	generated responses (eg error pages) needing to display the proxy
```
Comment (rousskov) [inline src/cf.data.pre]: s/eg/e.g.,/ to use customary spelling and for consistency sake

File: src/cf.data.pre [hunk 3]
```diff
@@ -7112,11 +7112,22 @@ LOC: Config.visibleHostname
...
 DOC_START
-	If you want to present a special hostname in error messages, etc,
-	define this.  Otherwise, the return value of gethostname()
-	will be used. If you have multiple caches in a cluster and
-	get errors about IP-forwarding you must set them to have individual
-	names with this setting.
+	The public domain name (FQDN) of this Proxy.
+
+	This name will be used in generated URLs sent to clients for
+	objects they need to download directly from Squid, and for other
```
Comment (rousskov) [inline src/cf.data.pre]: It would be good to add an example to the first part of the sentence. For example: `objects they need to download directly from Squid (e.g., style sheets for error pages and FTP icons)`.

Comment (rousskov) [inline src/cf.data.pre]: s/Squid, and/Squid and/ because this `and` does not separate independent clauses (and there are only two items total, so no Oxford comma either).

Comment (rousskov) [inline src/cf.data.pre]: AFAICT, `and for` appears to tie the second item to the wrong place in the beginning of the sentence (`for objects and for other generated responses needing to display` instead of the intended `in generated URLs and for other generated responses needing to display`. IMHO, the intent of the second item would be better relayed by something like `and to name this proxy in generated responses (e.g., error pages)`.

I do not insist on this change.

File: src/cf.data.pre [hunk 4]
```diff
@@ -7112,11 +7112,22 @@ LOC: Config.visibleHostname
...
+	generated responses (eg error pages) needing to display the proxy
+	name.
+
+	If not set, auto-detection of the system host name will be performed
+	based on the following details (in this order):
+		- rDNS name of first listening port address
+		- the return value of gethostname()
+		- the fallback name of "localhost"
```
Comment (rousskov) [inline src/cf.data.pre]: I would drop the `the fallback name of` part because 

* All these names are "fallback" names.
* After `rDNS name of port`, it could be misread as if we are computing the name of localhost (whatever that is) instead of just using the literal string `localhost`.

File: src/cf.data.pre [hunk 5]
```diff
@@ -7112,11 +7112,22 @@ LOC: Config.visibleHostname
...
+
+	This name will be used in generated URLs sent to clients for
+	objects they need to download directly from Squid, and for other
+	generated responses (eg error pages) needing to display the proxy
+	name.
+
+	If not set, auto-detection of the system host name will be performed
+	based on the following details (in this order):
+		- rDNS name of first listening port address
+		- the return value of gethostname()
```
Comment (rousskov) [inline src/cf.data.pre]: I would say `the name returned by gethostbyname(2)` because `the return value of gethostbyname()` is, technically, an integer. I do not insist on this change.

File: src/cf.data.pre [hunk 6]
```diff
@@ -7112,11 +7112,22 @@ LOC: Config.visibleHostname
...
+	The public domain name (FQDN) of this Proxy.
+
+	This name will be used in generated URLs sent to clients for
+	objects they need to download directly from Squid, and for other
+	generated responses (eg error pages) needing to display the proxy
+	name.
+
+	If not set, auto-detection of the system host name will be performed
+	based on the following details (in this order):
+		- rDNS name of first listening port address
```
Comment (rousskov) [inline src/cf.data.pre]: s/first/the first/

File: src/cf.data.pre [hunk 7]
```diff
@@ -7112,11 +7112,22 @@ LOC: Config.visibleHostname
...
+		- the return value of gethostname()
+		- the fallback name of "localhost"
+
+	If you have multiple caches in a cluster and get errors about
+	IP-forwarding you must set them to have individual names with
```
Comment (rousskov) [inline src/cf.data.pre]: s/you must/, you must/

Comment (rousskov) [inline src/cf.data.pre]: It is a bit difficult to correctly attribute/dereference `them` in this sentence. It is not quite `multiple caches` nor it is `errors` mentioned earlier. I suggest `configure each cache with a unique visible_hostname (or unique_hostname) value`. I do not insist on this change.

File: src/cf.data.pre [hunk 8]
```diff
@@ -7112,11 +7112,22 @@ LOC: Config.visibleHostname
...
+	generated responses (eg error pages) needing to display the proxy
+	name.
+
+	If not set, auto-detection of the system host name will be performed
+	based on the following details (in this order):
+		- rDNS name of first listening port address
+		- the return value of gethostname()
+		- the fallback name of "localhost"
+
+	If you have multiple caches in a cluster and get errors about
```
Comment (rousskov) [inline src/cf.data.pre]: s/caches/proxies/? I do not insist on this change.

File: src/cf.data.pre (hunk 1 above)
Comment (yadij) [reply src/cf.data.pre]: The name configured here does get used in messages going out into the public Internet, so admin wishing to keep their internal naming scheme private need to be made aware of that and set this to something that is suitable for public display.

We do see a lot of configs using internal naming and complaining that Squid publishes this detail - if they set it appropriately to begin with things like Via would not be such a big privacy/security "problem".

Comment (yadij) [reply src/cf.data.pre]: done

File: src/cf.data.pre (hunk 2 above)
Comment (yadij) [reply src/cf.data.pre]: done

File: src/cf.data.pre (hunk 3 above)
Comment (yadij) [reply src/cf.data.pre]: done

Comment (yadij) [reply src/cf.data.pre]: resolved by the other change to this sentence

Comment (yadij) [reply src/cf.data.pre]: resolved by the other changes to this sentence.

File: src/cf.data.pre (hunk 5 above)
Comment (yadij) [reply src/cf.data.pre]: done

File: src/cf.data.pre (hunk 6 above)
Comment (yadij) [reply src/cf.data.pre]: done

File: src/cf.data.pre (hunk 8 above)
Comment (yadij) [reply src/cf.data.pre]: done

File: src/cf.data.pre (hunk 7 above)
Comment (yadij) [reply src/cf.data.pre]: re-phrased the sentence. Is the new version agreeable for you?

Comment (yadij) [reply src/cf.data.pre]: done

File: src/cf.data.pre (hunk 4 above)
Comment (yadij) [reply src/cf.data.pre]: done

File: src/cf.data.pre (hunk 1 above)
Comment (rousskov) [reply src/cf.data.pre]: > The name configured here does get used in messages going out into the public Internet

If you want to warn the admins about leaking naming information to the "public internet", please do so, but "public domain name of this Proxy" conveys a completely different message to many (probably most) readers.

Besides, not all proxies communicate with the public internet.

I continue to recommend describing what this name is used for and letting the admin decide what name they want to use for those purposes. Based on the other updated text, calling this name "host name" or "domain name" is sufficient for this context.

File: src/cf.data.pre (hunk 7 above)
Comment (rousskov) [reply src/cf.data.pre]: The `set each proxy to have individual names with this directive` phrase is still hard to comprehend. The expressions "to set a proxy to have something" and "to set proxy with this directive" are awkward/unusual; "individual" does not mean unique. To me, the phrase reads like poor automated translation from another language.

I do not insist on polishing this sentence further, but I still suggest using "unique" instead of "individual" and avoiding "to set a proxy" expression. FWIW, I can give you two more sketches to work with:

* give each proxy a unique name using a visible_hostname or unique_hostname directive
* uniquely name each proxy using a visible_hostname or unique_hostname directive

File: src/cf.data.pre (hunk 3 above)
Comment (rousskov) [reply src/cf.data.pre]: Not resolved. Exactly the same "for foo and for bar" problem remains.

File: src/cf.data.pre (hunk 2 above)
Comment (rousskov) [reply src/cf.data.pre]: You missed a comma after `e.g.` in at least one cases. The rules about following `e.g.` with a comma differ from culture to culture, but we should use a comma for consistency sake, at least when there are multiple examples (although I would just use it everywhere for simplicity).

```
$ fgrep 'e.g.,' -c src/cf.data.pre 
28
$ grep 'e[.]g[.][^,]' -c src/cf.data.pre 
6 # all for single-item examples
```

Besides, using a comma is the more logical than not using it because "for example" is followed by a comma and that is what "e.g." means :\-).

File: src/cf.data.pre (hunk 7 above)
Comment (yadij) [reply src/cf.data.pre]: This is getting more obtuse and unique_hostname already explains the situation. So replacing the whole paragraph with a far simpler "see also ..." reference.

**Comment by yadij** [comment]
@rousskov, can you review the new texts please? I'd like to get the downstream bugs abut this closed next months release.

File: src/cf.data.pre [hunk 9]
```diff
@@ -7112,11 +7112,28 @@ LOC: Config.visibleHostname
...
+	over the Internet are required by RFC 1123 to have a name which
+	can be resolved in DNS (an FQDN).
+
+	If not set, auto-detection of the system host name will be performed
+	based on the following details (in this order):
+		- rDNS name for the first listening port address
+		- the name returned by gethostname(2)
+		- the name "localhost"
+
+	This name will be used in generated responses sent to clients
```
Comment (rousskov) [inline src/cf.data.pre]: This sentence is very difficult to understand. `objects they need to display the proxy name` just does not compute. I suggest something like `This directive is used for naming this proxy in generated responses sent to clients. For example, it is used to expand %h in a Squid error page template`. Please use the correct `%code` if you follow this sketch -- I did not check whether `%h` is correct.

File: src/cf.data.pre [hunk 10]
```diff
@@ -7112,11 +7112,28 @@ LOC: Config.visibleHostname
...
+	for objects they need to display the proxy name. For example;
+	error pages.
+
+	This name will also be used in generated URLs sent to clients for
+	objects they need to download directly from Squid. For example;
+	CSS stylesheets, FTP icons.
+
+	See unique_hostname directive if you have multiple proxies in a
+	cluster which need to share a visible hostname.
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/cf.data.pre]: The subject here is a plural 'proxies'. Plural of "that" is "those" - so: "proxies in a cluster those need to share a". Does not make sense.

File: src/cf.data.pre (hunk 9 above)
Comment (yadij) [reply src/cf.data.pre]: You want all of the past present and future codes to be listed as well as all the non-errorpage uses? error pages naming the proxy is just the most obvious and best-known example.

File: src/cf.data.pre [hunk 11]
```diff
@@ -7112,11 +7112,28 @@ LOC: Config.visibleHostname
...
 DEFAULT_DOC: Automatically detect the system host name
 DOC_START
-	If you want to present a special hostname in error messages, etc,
-	define this.  Otherwise, the return value of gethostname()
-	will be used. If you have multiple caches in a cluster and
-	get errors about IP-forwarding you must set them to have individual
-	names with this setting.
+	The hostname of this proxy.
+
+	Hosts connected to the Internet or handling messages which travel
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/cf.data.pre]: This paragraph sets the context for behaviour described in the following paragraphs. Removing or placing it last makes all this seems like arbitrary decisions we are imposing on admin ourselves. Removing it also perpetuates the admin misbelief that they can safely set it to random strings.

File: src/cf.data.pre (hunk 7 above)
Comment (yadij) [reply src/cf.data.pre]: resolving as this is superceded by the latest changes.

File: src/cf.data.pre (hunk 10 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/cf.data.pre]: > Plural of "that" is "those"

... when used as a [demonstrative](https://en.wikipedia.org/wiki/Demonstrative). In our context, `that` is not a demonstrative. It is a pronoun that introduces a relative [clause](https://en.wikipedia.org/wiki/English_relative_clauses). Unlike demonstratives, in our context, `that` is a part of speech that does not have singular and plural forms at all -- that is why you intuitively know that `proxies those need to share` simply "does not make sense" here.

The reason we should use `that` instead of `which` in our context is because the word in question introduces a [_restrictive_](https://en.wikipedia.org/wiki/English_relative_clauses#Restrictive_or_non-restrictive_relative_clauses) clause rather than a non-restrictive one. Restrictive clauses [should](https://www.grammarly.com/blog/which-vs-that/) be introduced by `that`.

File: src/cf.data.pre (hunk 11 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/cf.data.pre]: > The whole point of this directive is to change the name of the host where Squid runs as seen by external/remote/public entities.

I disagree. Those external/remote/public entities see a proxy service first and a host (hosting a proxy service) second (or not at all). In fact, in many contexts, when addressing the proxy service by the name configured here, those entities will not talk to the host where this Squid is running!

> This paragraph sets the context for behaviour described in the following paragraphs.

I am not sure it does. Naming is needed not because some RFC requires hosts to be named. Naming is needed for the use cases we describe further below. The first/previous paragraph is enough to set the general context IMO.

> Removing or placing it last makes all this seems like arbitrary decisions we are imposing on admin ourselves. 

I disagree. We document where this name is used. There is nothing "arbitrary" about those locations -- they are all driven by the same need to name "this proxy" or "this set of proxies"  for addressing it via HTTP and other relevant protocols. The admin should be able to make this naming decision based on the subset of the applicable-to-the-admin use cases (and the sited RFC would be of little, if any, help to that admin).

File: src/cf.data.pre (hunk 1 above)
Comment (yadij) [reply src/cf.data.pre]: > > The name configured here does get used in messages going out into the public Internet
> 
> If you want to warn the admins about leaking naming information to the "public internet", please do so, but "public domain name of this Proxy" conveys a completely different message to many (probably most) readers.
> 
> Besides, not all proxies communicate with the public internet.

Irrelevant. Most do and this directive is implemented for use on Internet.

> 
> I continue to recommend describing what this name is used for and letting the admin decide what name they want to use for those purposes.

This directive is used as an FQDN in URLs (as documented already). Other options use this name as a default value. Admin can set those other options if they have specific needs in those other use-cases.


> Based on the other updated text, calling this name "host name" or "domain name" is sufficient for this context.

That you use "host name" indicates a failure of clarity and why I am trying to be explicit about its FQDN nature. Setting it to a "host name" is a malformed config which causes invalid URI (possibly dangerous) to be transmitted. It can be a "hostname" (note lack of space in the term) - which is a valid FQDN in the networks without public DNS or traffic.

Comment (rousskov) [reply src/cf.data.pre]: > Alex: not all proxies communicate with the public internet.

> Amos: Irrelevant. Most do and this directive is implemented for use on Internet.

The fact that some proxies (that correctly use this directive) do not communicate with the public internet is certainly relevant to the directive description. When a directive can be successfully/correctly used for A and B, we should not describe it as "used for A". I cannot accept the "implemented for A [and not B]" argument (if that is what you are implying) either -- in this context, it is not very important what specific use cases this directive was implemented for (and I very much doubt it was implemented to exclude private network use cases).

> Alex: I continue to recommend describing what this name is used for and letting the admin decide what name they want to use for those purposes.

> Amos: This directive is used as an FQDN in URLs

Based on your statements quoted further below, I assume that here you count dotless hostnames in private networks as FQDNs. If so, you may be technically correct (you did not substantiate your claims, and I did not check myself), but that is not how most readers would interpret the above phrase. We should not use misleading (to most readers) phrasing in a directive documentation even if it that phrasing is technically accurate.

> Other options use this name as a default value. Admin can set those other options if they have specific needs in those other use-cases.

Yes, of course. IMO, it would be good to reference those other options in this directive documentation. I found two directives that explicitly refer to visible_hostname (unique_hostname and httpd_accel_surrogate_id) but perhaps there are more. If it is too difficult to provide a comprehensive list, we can just warn the reader that there are some undocumented dependencies. This is not what _this_ change request is about though.

> That you use "host name" indicates a failure of clarity and why I am trying to be explicit about its FQDN nature ... It can be a "hostname" (note lack of space in the term) - which is a valid FQDN in the networks without public DNS

Using "hostname" (without the space) is fine with me! And if you really want to add some text explaining why this hostname should be an FQDN, please feel free do so. Just assume that most readers would interpret FQDN as "a domain name ending with a TLD" or, worse, "a domain name with dots in it". IMHO, squid.conf.documented is not the place to educate readers about other FQDN meanings. We should not expect them to know those other meanings. And we should not assume they actually expect those other meanings in basic squid.conf.documented phrases even if they know about those other meanings.

To summarize, let's drop "public" and "FQDN" from the brief directive description. Let's document how this hostname is used by Squid. Adding followup paragraphs highlighting the dangers of not using an FQDNs for this parameter, especially on public networks, is perfectly fine, of course.

Comment (rousskov) [reply src/cf.data.pre]: Err... This is actually a thread in an old review made obsolete but the later [review](https://github.com/squid-cache/squid/pull/302#pullrequestreview-173094053) and branch changes. I do not know why you resumed this argument. FWIW, the current `The hostname of this proxy` brief description is fine with me and has been so since Nov 8, 2018!

Comment (yadij) [reply src/cf.data.pre]: Was working through the open discussions. Resolving this one now since current text is fine.
